Sprint 12

Kolesnikov-Eugene opened the pull request.

Kolesnikov-Eugene committed:
ImagesListService, PhotoResult and Photo struct added to the project

Kolesnikov-Eugene committed:
nextPage count added

Kolesnikov-Eugene committed:
imageFeed downloads images from server.
SingleImage is downloaded from server.

Kolesnikov-Eugene committed:
AlertPresenter and AlertModel added.
Change like functionality added.

Kolesnikov-Eugene committed:
Full functionality:
- cash cleaning method
- AlertPresenter and AlertModel

Kolesnikov-Eugene committed:
CookiesCleaner class to clean cookies
Alert on exit from profile
code refactoring

Kolesnikov-Eugene committed:
final ver

Helvetii951 commented on FunImages/Alert/AlertPresenter.swift:
```diff
@@ -0,0 +1,33 @@
+        if alertHasTwoButtons {
+            guard let text = model.cancelButtonText else { return }
+            let cancelAction = UIAlertAction(title: text, style: .default) { _ in
+            }
```
Здесь можно убрать блок с действием, он все-равно ничего не делает.

Helvetii951 commented on FunImages/AppDelegate.swift:
```diff
@@ -6,14 +6,17 @@
-        // Override point for customization after application launch.
+        ProgressHUD.animationType = .systemActivityIndicator
+        ProgressHUD.colorHUD = .black
+        ProgressHUD.colorAnimation = .lightGray
```
Рекомендую для подобных настроек сделать отдельный приватную функцию.

Helvetii951 commented on FunImages/Helpers/CookiesCleaner.swift:
```diff
@@ -0,0 +1,23 @@
+import Foundation
+import WebKit
+
+final class CookiesCleaner {
+    static func cleanCookies() {
+        HTTPCookieStorage.shared.removeCookies(since: Date.distantPast)
+        
+        WKWebsiteDataStore.default().fetchDataRecords(ofTypes: WKWebsiteDataStore.allWebsiteDataTypes()) { records in
+            records.forEach { record in
+                WKWebsiteDataStore.default().removeData(ofTypes: record.dataTypes, for: [record], completionHandler: {})
+            }
+        }
+    }
+    
+    private init() { }
+}
```
Лучше убрать этот инит. В нем нет смысла.

Helvetii951 commented on FunImages/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,137 @@
+                switch result {
+                case .success(let photoResult):
+                    self.task = nil
+                    photos += photoResult.map { Photo(
+                        id: $0.id,
+                        size: CGSize(width: Double($0.width), height: Double($0.height)),
+                        createdAt: self.dateFormatter.date(from: $0.createdAt),
```
За использование map 👍

Helvetii951 commented on FunImages/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,137 @@
+                switch result {
+                case .success(let photoResult):
+                    self.task = nil
+                    photos += photoResult.map { Photo(
+                        id: $0.id,
+                        size: CGSize(width: Double($0.width), height: Double($0.height)),
+                        createdAt: self.dateFormatter.date(from: $0.createdAt),
+                        welcomeDescription: $0.description,
+                        thumbImageUrl: $0.urls.thumb,
+                        largeImageUrl: $0.urls.full,
+                        isLiked: $0.likedByUser
+                    )
+                    }
```
Здесь форматирование немного некорректно выглядит. Лучше сделать перенос строки перед `Photo(.

Helvetii951 commented on FunImages/ImagesList/ImagesListService.swift:
```diff
@@ -0,0 +1,137 @@
+            likeTask?.cancel()
+        }
+        
+        var httpMethodForLike = "DELETE"
+        if !isLiked {
+            httpMethodForLike = "POST"
+        }
```
Тут можно упростить с помощью [тернарного оператора](https://www.programiz.com/swift-programming/ternary-conditional-operator).
```
let httpMethodForLike = isLiked ? "POST": "DELETE"
```

Helvetii951 commented on FunImages/Profile/ProfileViewController.swift:
```diff
@@ -137,19 +137,33 @@ final class ProfileViewController: UIViewController {
+                    }
                 }
+                
+                self.present(SplashViewController(), animated: true)
```
Здесь ошибка. Нужно менять rootViewController окна приложения на SplashViewController. Сейчас он презентуется.

Helvetii951 commented on FunImages/Profile/ProfileViewController.swift:
```diff
@@ -137,19 +137,33 @@ final class ProfileViewController: UIViewController {
+                
+                CookiesCleaner.cleanCookies()
+                
+                for view in self.view.subviews {
+                    if view is UILabel {
+                        view.removeFromSuperview()
+                    }
+                    else {
+                        if let view = view as? UIImageView {
+                            view.image = UIImage(named: "profile_photo_placeholder")
+                        }
+                    }
                 }
```
Не совсем понятно зачем эти манипуляции с элементами UI.

Helvetii951 commented on FunImages/Profile/ProfileViewController.swift:
```diff
@@ -137,19 +137,33 @@ final class ProfileViewController: UIViewController {
+            message: "Уверены, что хотите выйти?",
+            okButtonText: "Да",
+            cancelButtonText: "Нет") {
+                OAuth2TokenStorage.shared.deleteToken()
+                
+                CookiesCleaner.cleanCookies()
```
Желательно объединять вынести эти действия в отдельный метод.

Helvetii951 commented on FunImages/SingleImage/SingleImageViewController.swift:
```diff
@@ -6,38 +6,74 @@
     @IBAction func didTapShareButton(_ sender: UIButton) {
         let activityController = UIActivityViewController(
-            activityItems: [image!],
+            activityItems: [bigImage!],
```
Очень не рекомендую использовать force unwrap. Может привести к неприятным падениям. Лучше обработать через guard, и не вызывать `UIActivityViewController`, если `bigImage  == nil`

Helvetii951 commented on FunImages/ImagesList/ImagesListViewController.swift:
```diff
@@ -22,42 +25,62 @@ final class ImagesListViewController: UIViewController {
+                        return
+                    }
+                    self.updateTableViewAnimated()
+                })
```
Здесь необязательно распаковывать self, можно использовать optional chaining
```
self?.updateTableViewAnimated()
```

Kolesnikov-Eugene committed:
added func configureProgressHUD to AppDelegate
private init deleted from CookiesCleaner
func getDateToDisplay added to ImagesListViewController to return empty string in case if there is no date to show from server
ProfileImage rounded.
func didTapLogOutButton changed (rootViewController changing to SplashViewController).
